widget/embedded/src/utils/colors.ts

```diff
@@ -1,141 +1,159 @@
+  green: number;
+  blue: number;
+};
 
-export function generateShades(
-  baseColor: string,
-  prefix: string,
-  reverse?: boolean
-): { [key: string]: string } {
-  const BASE_SHADE_INDEX = 100;
-  const NUMBER_OF_COLORS = 9;
-  const SHADING_INCREMENT_MAX = 10; // For bright colors
-  const SHADING_INCREMENT_MIN = 6; // For dark colors
-
-  const HALF_NUMBER_OF_COLORS = 4;
-
-  const shadeIncrements: number[] = [];
-  let increment = 90;
-  for (let i = 0; i < NUMBER_OF_COLORS; i++) {
-    shadeIncrements.push(increment);
-    if (i + 1 === HALF_NUMBER_OF_COLORS) {
-      increment = 0;
-    } else {
-      increment -=
-        i >= HALF_NUMBER_OF_COLORS
-          ? SHADING_INCREMENT_MIN
-          : SHADING_INCREMENT_MAX;
-    }
-  }
+type ShadeOrTintFunction = (rgb: RGB, step: number) => RGB;
 
-  // If 'reverse' is true, that means the color should be from dark to light
-  const decrementAmounts = reverse
-    ? shadeIncrements.slice().reverse()
-    : shadeIncrements;
+const HEX_BASE = 16;
+const MAX_COLOR_VALUE = 255;
+const PERCENT = 0.11;
 
-  const shades: { [key: string]: string } = {};
+// Utility Functions
 
-  // Generate shades
-  for (let i = 0; i < NUMBER_OF_COLORS; i++) {
-    const shadeIndex = BASE_SHADE_INDEX + i * BASE_SHADE_INDEX;
-    const decrementAmount = decrementAmounts[i];
+/*
+ * We letting users to override some specific colors (e.g. `primary550`, `secondary100`).
+ * So we are generating a range of colors if `primary` (or other keys) has passed but if user is passing a specific color,
+ * we will override the user color to generated range.
+ */
+function isOverridingColor(colorKey: string): boolean {
+  return /[0-9]+$/.test(colorKey);
+}
 
-    const shade: { [key: string]: string } = {
-      [`${prefix}${shadeIndex}`]: generateShade(baseColor, decrementAmount),
-    };
-    Object.assign(shades, shade);
-  }
+// pad a hexadecimal string with zeros if it needs it
+function pad(number: string, length: number) {
+  return number.padStart(length, '0');
+}
 
-  return { ...shades, [prefix]: baseColor };
+// Convert an integer to a 2-character hex string
+function intToHex(value: number) {
+  const clampedValue = Math.min(
+    Math.max(Math.round(value), 0),
+    MAX_COLOR_VALUE
+  );
+  return pad(clampedValue.toString(HEX_BASE), 2);
 }
 
-export function generateShade(
-  baseColor: string,
-  decrementAmount: number
-): string {
-  const [r, g, b] = convertToRGB(baseColor);
-  return `#${clamp(r + decrementAmount)}${clamp(g + decrementAmount)}${clamp(
-    b + decrementAmount
-  )}`;
+// convert one of our rgb color objects to a full hex color string
+function rgbToHex(rgb: RGB) {
+  return `#${intToHex(rgb.red)}${intToHex(rgb.green)}${intToHex(rgb.blue)}`;
 }
 
-export function convertToRGB(baseColor: string): number[] {
-  const COLOR_SLICE_START = 1;
-  const COLOR_SLICE_END = 3;
-  const BASE_COLOR_LENGTH = 7;
+// Convert a hex color string to an RGB object
+function hexToRGB(hex: string): RGB {
```

Comment:
You are handling 6 digit hex colors but 3 digit hex colors ae not handled here. Therefore, if the secondary color is set to something like `#129`, it won't work properly.